File: python-version/subscription.py
from datetime import datetime, timedelta

class Subscription:
    def __init__(self, sub_id, software_name, license_key, start_date, duration, price):
        self.sub_id = sub_id
        self.software_name = software_name
        self.license_key = license_key
        self.start_date = start_date
        self.expiry_date = start_date + timedelta(days=duration)
        self.price = price
        self.renewal_history = []

    def renew(self, duration):
        if self.expiry_date < datetime.now():
            self.expiry_date = datetime.now() + timedelta(days=duration)
        else:
            self.expiry_date += timedelta(days=duration)
        self.renewal_history.append((datetime.now(), duration))

File: java-version/Subscription.java
// File: Subscription.java
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;

public class Subscription {
    private int subId;
    private String softwareName;
    private String licenseKey;
    private LocalDate startDate;
    private LocalDate expiryDate;
    private double price;
    private List<String> renewalHistory;

    public Subscription(int subId, String softwareName, String licenseKey, LocalDate startDate, int durationDays, double price) {
        this.subId = subId;
        this.softwareName = softwareName;
        this.licenseKey = licenseKey;
        this.startDate = startDate;
        this.expiryDate = startDate.plusDays(durationDays);
        this.price = price;
        this.renewalHistory = new ArrayList<>();
    }

    public void renew(int additionalDays) {
        renewalHistory.add("Renewed on " + LocalDate.now() + " for " + additionalDays + " days");
        expiryDate = expiryDate.plusDays(additionalDays);
    }

    public LocalDate getExpiryDate() {
        return expiryDate;
    }

    public String getSoftwareName() {
        return softwareName;
    }
}
